alacritty_terminal/src/input.rs

```diff
@@ -903,55 +902,42 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
         *self.ctx.received_count() += 1;
     }
 
-    /// Attempts to find a binding and execute its action
+    /// Attempt to find a binding and execute its action.
     ///
     /// The provided mode, mods, and key must match what is allowed by a binding
     /// for its action to be executed.
     ///
-    /// Returns true if an action is executed.
-    fn process_key_bindings(&mut self, input: KeyboardInput) -> bool {
-        let mut has_binding = false;
-        for binding in self.key_bindings {
-            let is_triggered = match binding.trigger {
-                Key::Scancode(_) => binding.is_triggered_by(
-                    *self.ctx.terminal().mode(),
-                    input.modifiers,
-                    &Key::Scancode(input.scancode),
-                    false,
-                ),
-                _ => {
-                    if let Some(key) = input.virtual_keycode {
-                        let key = Key::from_glutin_input(key);
-                        binding.is_triggered_by(
-                            *self.ctx.terminal().mode(),
-                            input.modifiers,
-                            &key,
-                            false,
-                        )
-                    } else {
-                        false
-                    }
-                },
-            };
+    /// Return true if chars should be suppressed.
+    fn process_key_bindings(&mut self, input: KeyboardInput) {
+        let mode = *self.ctx.terminal().mode();
+
+        *self.ctx.suppress_chars() = self
+            .key_bindings
+            .iter()
+            .filter(|binding| {
+                let key = match (binding.trigger, input.virtual_keycode) {
+                    (Key::Scancode(_), _) => Key::Scancode(input.scancode),
+                    (_, Some(key)) => Key::from_glutin_input(key),
+                    _ => return false,
+                };
 
-            if is_triggered {
-                // binding was triggered; run the action
+                binding.is_triggered_by(mode, input.modifiers, &key, false)
+            })
+            .fold(None, |suppress_chars, binding| {
+                // Binding was triggered; run the action
                 binding.execute(&mut self.ctx, false);
-                has_binding = true;
-            }
-        }
 
-        has_binding
+                // Don't suppress when there has been a `ReceiveChar` action
+                Some(suppress_chars.unwrap_or(true) && binding.action != Action::ReceiveChar)
+            })
+            .unwrap_or(false); // Don't suppress char if no bindings were triggered
```

Comment:
Putting this comment above the `unwrap_or` on the same indentation level as the `.` should hopefully make things a little easier to read.

I'm pretty sure I've tested that and `rustfmt` handles that very well (rustfmt really works well).